lib/azure/armrest/virtual_machine_image_service.rb

```diff
@@ -31,23 +31,28 @@ def initialize(configuration, options = {})
       def list_all(location = @location)
         raise ArgumentError, "No location specified" unless location
 
+        mutex  = Mutex.new
         images = []
-        publishers(location).each do |publisher|
-          offers(location, publisher.name).each do |offer|
-            skus(offer.name, location, publisher.name).each do |sku|
-              versions(sku.name, offer.name, location, publisher.name).each do |version|
-                images << Azure::Armrest::VirtualMachineImage.new(
+
+        Parallel.each(publishers(location), :in_threads => configuration.max_threads) do |publisher|
+          Parallel.each(offers(location, publisher.name), :in_threads => configuration.max_threads) do |offer|
+            Parallel.each(skus(offer.name, location, publisher.name), :in_threads => configuration.max_threads) do |sku|
+              Parallel.each(versions(sku.name, offer.name, location, publisher.name), :in_threads => configuration.max_threads) do |version|
+                mutex.synchronize do
+                  images << Azure::Armrest::VirtualMachineImage.new(
```

Comment:
Where is the place that the code needs to make API calls to the provider? Is it in the constructor? Maybe we can use multiple loops to collect the hashes, and finally use one `Parallel.each` to construct images.

Reply:
The `versions`, `skus` and `offers` and `publishers` methods are where the actual API calls happen.